kolibri/plugins/coach/class_summary_api.py

```diff
@@ -243,7 +230,25 @@ def retrieve(self, request, pk):
             exam_node_ids = [question['exercise_id'] for question in exam.get("question_sources")]
             all_node_ids |= set(exam_node_ids)
 
+        # map node ids => content_ids so we can replace missing nodes, if another matching content_id node exists
+        content_id_map = {resource['contentnode_id']: resource['content_id'] for lesson in query_lesson for resource in lesson.resources}
         query_content = ContentNode.objects.filter(id__in=all_node_ids)
+        # final list of available nodes
+        list_of_ids = [node.id for node in query_content]
```

Comment:
does this end up running the query twice? (Once here and once on line 263)

If not, this looks good.

If so, you can do something like

```python
content_data = data(ContentSerializer, query_content)
list_of_ids = [content['content_id'] for content in content_data]
```

and then below

```
-            "content": data(ContentSerializer, query_content),
+            "content": content_data,
```

Reply:
I believe it is cached.
https://docs.djangoproject.com/en/1.11/topics/db/queries/#caching-and-querysets